828 MCAP specification for osi tracefiles (#833)
OpenSimulationInterface/open-simulation-interface: doc/architecture/trace_file_mcap_format.adoc

ifndef::include-only-once[]
:root-path: ../
include::{root-path}_config.adoc[]
endif::[]
= MCAP Format

== General Requirements
- Must comply with the https://mcap.dev/spec[MCAP format specification] version `0x30`
- Must allow other non-OSI data to be present in the MCAP file
- Message records must be written into `chunk records` for indexed files
- Only OSI top-level messages containing a timestamp field are permitted to be directly stored in MCAP channels

> - Must contain only a single scenario with a unique global time

Reviewer: I remember that we talked about a common/unified time frame but not about that an MCAP file should only be allowed to contain a single scenario.

Also, I thought that the term "unique global time" is maybe a bit confusing as well. I'd rather write something like common time frame.

> - Must contain only a single scenario with a unique global time

Author: ```suggestion
- Must contain only a single scenario with a common time frame
```

> - Must contain only a single scenario with a unique global time

Reviewer: Still, I don't agree that an mcap file must contain only a single scenario. I think the word "scenario" could be misleading. You can put an arbitrary number of scenarios in one trace file. IMO the only relevant information here is that all the messages in the file must have a common time frame.

> - Must contain only a single scenario with a unique global time

Author: ```suggestion
- Must be limited to a single, unified sequence of events within the same time frame.
```

- An MCAP file is considered a single dataset

== Schema

> - `name` field: Full message type name, including package (e.g., `osi3.SensorData`)

Author: I understand the [protobuf documentation](https://protobuf.dev/reference/cpp/api-docs/google.protobuf.descriptor/#Descriptor.full_name.details) as if this is defined.  But to be more precise I will change it to  `fully-qualified name of the protobuf message type`

> Especially, because you used "OSI3::SensorData" in line 60 (not in the same context but anyways).

There I meant the [osi3::SensorData Struct](https://opensimulationinterface.github.io/osi-antora-generator/asamosi/latest/gen/structosi3_1_1SensorData.html). But it makes sense to simply write  "A channel containing OSI `SensorData` messages" and circumvent this. I will change line 60

- `encoding` field: Must be `protobuf`
- `data` field: String-encoded `google::protobuf::FileDescriptorSet` for the OSI top-level message

== Channel
- `message_encoding` field: Must be `protobuf``
- `metadata` field:
** Must include an `osi_version` key, specifying the OSI SemVer version of the OSI top-level message contained within the channel
** Must include a `protobuf_version` key, specifying the protobuf SemVer version used to create the OSI top-level message contained within the channel
** Should include a `description` key, explaining the data's origin and purpose in natural language.

== Message
- `publish_time` field: 
** Must reflect the timestamp of the stored OSI top-level message
** Must be in nanoseconds

> - `log_time` field: Must reflect the time when the message was enqueued for MCAP file addition

Author: Can you please elaborate further?

** Must reflect the timestamp of the stored OSI top-level message
** Must be in nanoseconds

== File-wide Metadata

> - Must include metadata with the name `versions` containing at least the following key-value pair:

Author: I do not have any opinion about this. @jdsika had the idea to add prefixes, so lets see what he suggests.

** `osi`: SemVer version of the minimum required OSI version
- Must include metadata with the name `asam_osi` containing at least the following key-value pairs:
** `zero_time`: ISO 8601 YYYYMMDDThhmmss.f formatted point in time representing the zero time of the scenario

> ** `timestamp`: ISO 8601 YYYYMMDDThhmmss.f formatted creation time of the file 

Author: I would personally also favor the unix epoch in ms or nanoseconds. The proposal is using ISO 8601 as it seems like that was used for trace file names etc. before. 

That would also make more sense to be compliant with https://opensimulationinterface.github.io/osi-antora-generator/asamosi/latest/gen/structosi3_1_1EnvironmentalConditions.html#a636bb78627046f34208f42f586ab2086?

Lets wait an see if anyone disagrees.

> - It is strongly recommended to include metadata with the name `asam_osi` containing the following key-value pairs:

Reviewer: Should there be two "asam_osi" metadata records (see two lines above) or should all the metadata fields be in one metadata record?

Author: While I don't know if it is technically possible to have two metadata records with the same name (I assume it might be), it would make more sense to have all of that in one record. Considering your proposal to always speak of "a metadata record" in this context (see other comment) it should be clearer once this has been added.

** `description`: Short human-readable scenario description

> ** `creator`: csv of person or company (not tool) creating the file

Author: Yes I will write it out once.

> ** `license`` csv of spdx identifiers

Reviewer: I think SPDX also has options to combine multiple licenses (expressions), so I don't know if csv is even necessary here.

*** A file metadata in a new metadata category with the arbitrary name `specification` contains the key `openDrive-road-rule` with the value `RHT`
** Cycle time variation of a sensor
*** Assume you want to express the interface cycle time variation of a sensor.
*** The `context` category contains the key `iso_23150` with the value` ISO 23150:2011`
*** A channel containing `OSI3::SensorData` messages has metadata with the key `iso_23150-cycle-time-variation:` and the value `80`

== Compression
- OSI-compliant tooling must support compression types: `none`, `lz4`, and `zstd`

== Naming Convention
.mcap files must follow this naming convention:

----
<opt. prefix>_<opt. timestamp>_<type>_<opt. suffix>.mcap
----

When not using an optional field, the corresponding underscore delimiter must be omitted so that no double underscore is present.

[#tab-MCAP-file-naming-convention]
.MCAP file naming convention
[cols="1,1"]
|===
|Field |Explanation 

|opt. prefix
|An optional prefix which may be used to specify the type of scenario (e.g. `cut-in`) or uniqueness of the setup (e.g. `target-5m`). May not contain any `_` characters.

> |opt. timestamp

Reviewer: Maybe we should state that the timestamp can only exceptionally be omitted when there really is no reference to a global time in the file?

I think if you have a real-world capture or any other trace file that has any meaningful relation to a global time frame, it should be visible in the filename.

Reviewer 2: In my opinion the meaningful relation to a global time is the exception here, since simulation should be the "normal" use-case rather than measurements.

> |Defines the absolute start time for a scenario or recording. If following the recommended zero time for the timestamps of the top-level messages, this time must represent the zero time. The format must adhere to ISO 8601 [cite:iso8601].

Author: To my limited understanding of the ISO 8601, the .osi and .txth spec is not stating a format. I just provides the example `20210818T150542Z` and refers to ISO 8601 

Something like `YYYYMMDDThhmmssZ` would be a valid format ISO 8601 but  `YYYY-MM-DD-HH` as well right right? 

I would like to add the format  `YYYYMMDDThhmmssZ` and the mention that it must be in UTC (not global, due to the Z) to the .osi/.txth trace file naming convention but this assumes that generalizing the example is not considered a breaking change.

> |Defines the absolute start time for a scenario or recording. If following the recommended zero time for the timestamps of the top-level messages, this time must represent the zero time. The format must adhere to ISO 8601 [cite:iso8601].

Reviewer: The naming convention even states that the recommended format is YYYMMDDThhmmssZ (even though with exemplary numbers). I would just add the format specification and maybe even consider decimal places. The recommended format remains the same as far as I'm concerned.

|type
| Specifies the type of the contained the top-level message(s) and must be one of the following values:

`sv` file contains only `SensorView` messages. +
`gt` file contains only `GroundTruth` messages. +
`hvd` file contains only `HostVehicleData` messages. +
`sd` file contains only SensorData` messages. +
`tc` file contains only `TrafficCommand` messages. +
`tcu` file contains only `TrafficCommandUpdate` messages. +
`tu` file contains only `TrafficUpdate` messages. +
`mr` file contains only `MotionRequest` messages. +
`su` file contains only `StreamingUpdate` messages. +
`multi` file contains multiple, different types of of top-level messages (not including different channels of the same type).

|opt. suffix
|An optional suffix which may be used the same way as the optional prefix or be used to specify further details like the minimum required OSI version. May not contain any `_` characters.

|===

**Example**

The following list shows examples of valid OSI MCAP file names:

> - `20210818T150542Z_highway_sv.mcap`

Reviewer: Why is there "highway" between the timestamp and the type specification?

Author: ```suggestion
- `20210818T150542Z_sv_highway.mcap`
```

= MCAP Format

== General Requirements
- Must comply with the https://mcap.dev/spec[MCAP format specification] version `0x30`
- Must allow other non-OSI data to be present in the MCAP file
- Message records must be written into `chunk records` for indexed files

> 

Author: Add note explaining that only top-level messages with timestamp can be saved

== Schema
- `name` field: Full message type name, including package (e.g., `osi3.SensorData`)
- `encoding` field: Must be `protobuf`
- `data` field: String-encoded `google::protobuf::FileDescriptorSet` for the OSI top-level message

== Channel
- `message_encoding` field: Must be "protobuf"

> - `metadata` field: Should include a brief description of data origin and purpose

Author: Add OSI and protobuf version as metadata for each channel for easier debugging etc. with one file supporting channels with different versions.

Author: Done but only added osi version.

== Message
- `publish_time` field: 
  - Must reflect the timestamp of the stored OSI top-level message if it contains a timestamp field

>   - Should represent a plausible value to express message order in relation to timestamps of other top-level messages if the message doesn't contain a timestamp (e.g. SensorViewConfiguration)

Author: Remove this sentence for now as `SensorViewConfiguration` has no real timestamp, see  discussion in #392

Author: Done. Removed text.

  - Must be in nanoseconds
- `log_time` field: Must reflect the time when the message was enqueued for MCAP file addition
  - Must reflect the time when the top-level OSI message was enqueued for addition to the MCAP file
  - Zero time doesn't need to coincide with the Unix epoch, but it is recommended
  - Must be in nanoseconds

== Metadata
- Must include metadata with the name `versions` containing at least the following key-value pairs:

>   * `osi`: SemVer version of OSI specification

Reviewer 2: This shall be the highest compatible OSI version of all OSI messages in the MCAP file.

Author: Done. Precised text.

>   * `protobuf`: SemVer version of protobuf library

Reviewer 2: Do we need the protobuf version at all or can we remove them? Also from the file name specification of the .osi format.

Author: During today's meeting we agreed to store it the channels metadata (see comments below or updated files).  Thus resolving this thread.

- Must include metadata with the name `creation_date` containing at least the following key-value pairs:

>   * `timestamp`: ISO 8601 formatted creation time

Reviewer 2: Add description of the exact time this timestamp shall represent

- Must include metadata with the name `description` containing at least the following key-value pairs:
  * `text`: Scenario description
- Additional custom metadata may be added

== Compression
- OSI-compliant tooling must support compression types: `none`, `lz4`, and `zstd`

> - Recommended compression in use should be either `lz4` or `zstd`

Author: Remove sentence as utilities will support a default compression.
